File: apollo-integration/src/test/java/com/apollographql/apollo/ApolloInterceptorChainTest.java
package com.apollographql.apollo;


import android.support.annotation.NonNull;

import com.apollographql.apollo.api.Operation;
import com.apollographql.apollo.api.Response;
import com.apollographql.apollo.api.internal.Optional;
import com.apollographql.apollo.cache.normalized.Record;
import com.apollographql.apollo.exception.ApolloException;
import com.apollographql.apollo.integration.normalizer.EpisodeHeroNameQuery;
import com.apollographql.apollo.integration.normalizer.type.Episode;
import com.apollographql.apollo.interceptor.ApolloInterceptor;
import com.apollographql.apollo.interceptor.ApolloInterceptorChain;
import com.apollographql.apollo.interceptor.FetchOptions;
import com.apollographql.apollo.internal.interceptor.RealApolloInterceptorChain;

import org.junit.Assert;
import org.junit.Test;

import java.util.Collections;
import java.util.List;
import java.util.concurrent.Executor;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicInteger;

import javax.annotation.Nonnull;

import okhttp3.MediaType;
import okhttp3.Protocol;
import okhttp3.Request;
import okhttp3.ResponseBody;

import static com.apollographql.apollo.interceptor.ApolloInterceptor.CallBack;
import static com.apollographql.apollo.interceptor.ApolloInterceptor.InterceptorResponse;
import static com.google.common.truth.Truth.assertThat;

public class ApolloInterceptorChainTest {

  @Test
  public void onProceedAsyncCalled_chainPassesControlToInterceptor() throws TimeoutException, InterruptedException {
    final AtomicInteger counter = new AtomicInteger(1);

    EpisodeHeroNameQuery query = createQuery();

    ApolloInterceptor interceptor = new ApolloInterceptor() {
      @Override
      public void interceptAsync(@Nonnull InterceptorRequest request, @Nonnull ApolloInterceptorChain chain,
          @Nonnull Executor dispatcher, @Nonnull CallBack callBack) {
        counter.decrementAndGet();
      }

      @Override public void dispose() {

      }
    };

    List<ApolloInterceptor> interceptors = Collections.singletonList(interceptor);
    RealApolloInterceptorChain chain = new RealApolloInterceptorChain(interceptors);
    chain.proceedAsync(new ApolloInterceptor.InterceptorRequest(query, FetchOptions.NETWORK_ONLY, Optional.<Operation.Data>absent()),
        Utils.immediateExecutor(), new CallBack() {
          @Override public void onResponse(@Nonnull InterceptorResponse response) {

          }

          @Override public void onFailure(@Nonnull ApolloException e) {

          }

          @Override public void onCompleted() {

          }

          @Override public void onFetch(ApolloInterceptor.FetchSourceType sourceType) {

          }
        });

    //If counter's count doesn't go down to zero, it means interceptor's interceptAsync wasn't called
    //which means the test should fail.
    if (counter.get() != 0) {
      Assert.fail("Control not passed to the interceptor");
    }
  }

  @Test
  public void onProceedAsyncCalled_correctInterceptorResponseIsReceived() throws TimeoutException, InterruptedException {
    final AtomicInteger counter = new AtomicInteger(1);

    EpisodeHeroNameQuery query = createQuery();
    final InterceptorResponse expectedResponse = prepareInterceptorResponse(query);

    ApolloInterceptor interceptor = new ApolloInterceptor() {
      @Override
      public void interceptAsync(@Nonnull InterceptorRequest request, @Nonnull ApolloInterceptorChain chain,
          @Nonnull Executor dispatcher, @Nonnull final CallBack callBack) {
        dispatcher.execute(new Runnable() {
          @Override public void run() {
            callBack.onResponse(expectedResponse);
          }
        });
      }

      @Override public void dispose() {

      }
    };

    List<ApolloInterceptor> interceptors = Collections.singletonList(interceptor);
    RealApolloInterceptorChain chain = new RealApolloInterceptorChain(interceptors);
    chain.proceedAsync(new ApolloInterceptor.InterceptorRequest(query, FetchOptions.NETWORK_ONLY, Optional.<Operation.Data>absent()),
        Utils.immediateExecutor(), new CallBack() {
          @Override public void onResponse(@Nonnull InterceptorResponse response) {
            assertThat(response).isEqualTo(expectedResponse);
            counter.decrementAndGet();
          }

          @Override public void onFailure(@Nonnull ApolloException e) {

          }

          @Override public void onCompleted() {

          }

          @Override public void onFetch(ApolloInterceptor.FetchSourceType sourceType) {

          }
        });

    if (counter.get() != 0) {
      Assert.fail("Interceptor's response not received");
    }
  }

  @Test
  public void onProceedAsyncCalled_correctExceptionIsCaught() throws TimeoutException, InterruptedException {
    final AtomicInteger counter = new AtomicInteger(1);

    final String message = "ApolloException";
    EpisodeHeroNameQuery query = createQuery();
    ApolloInterceptor interceptor = new ApolloInterceptor() {
      @Override
      public void interceptAsync(@Nonnull InterceptorRequest request, @Nonnull ApolloInterceptorChain chain,
          @Nonnull Executor dispatcher, @Nonnull final CallBack callBack) {
        dispatcher.execute(new Runnable() {
          @Override public void run() {
            ApolloException apolloException = new ApolloException(message);
            callBack.onFailure(apolloException);
          }
        });
      }

      @Override public void dispose() {

      }
    };

    List<ApolloInterceptor> interceptors = Collections.singletonList(interceptor);
    RealApolloInterceptorChain chain = new RealApolloInterceptorChain(interceptors);
    chain.proceedAsync(new ApolloInterceptor.InterceptorRequest(query, FetchOptions.NETWORK_ONLY, Optional.<Operation.Data>absent()),
        Utils.immediateExecutor(), new CallBack() {
          @Override public void onResponse(@Nonnull InterceptorResponse response) {

          }

          @Override public void onFailure(@Nonnull ApolloException e) {
            assertThat(e.getMessage()).isEqualTo(message);
            counter.decrementAndGet();
          }

          @Override public void onCompleted() {

          }

          @Override public void onFetch(ApolloInterceptor.FetchSourceType sourceType) {

          }
        });

    if (counter.get() != 0) {
      Assert.fail("Exception thrown by Interceptor not caught");
    }
  }

  @Test
  public void onDisposeCalled_interceptorIsDisposed() {
    final AtomicInteger counter = new AtomicInteger(1);
    EpisodeHeroNameQuery query = createQuery();

    ApolloInterceptor interceptor = new ApolloInterceptor() {
      @Override
      public void interceptAsync(@Nonnull InterceptorRequest request, @Nonnull ApolloInterceptorChain chain, @Nonnull
          Executor dispatcher, @Nonnull CallBack callBack) {

      }

      @Override public void dispose() {
        counter.decrementAndGet();
      }
    };

    List<ApolloInterceptor> interceptors = Collections.singletonList(interceptor);
    RealApolloInterceptorChain chain = new RealApolloInterceptorChain(interceptors);
    chain.dispose();
    if (counter.get() != 0) {
      Assert.fail("Interceptor's dispose method not called");
    }
  }

  @NonNull
  private EpisodeHeroNameQuery createQuery() {
    return EpisodeHeroNameQuery
        .builder()
        .episode(Episode.EMPIRE)
        .build();
  }

  @NonNull
  private InterceptorResponse prepareInterceptorResponse(EpisodeHeroNameQuery query) {
    Request request = new Request.Builder()
        .url("https://localhost:8080/")
        .build();

    okhttp3.Response okHttpResponse = new okhttp3.Response.Builder()
        .request(request)
        .protocol(Protocol.HTTP_2)
        .code(200)
        .message("Intercepted")
        .body(ResponseBody.create(MediaType.parse("text/plain; charset=utf-8"), "fakeResponse"))
        .build();

    Response<EpisodeHeroNameQuery.Data> apolloResponse = Response.<EpisodeHeroNameQuery.Data>builder(query).build();

    return new InterceptorResponse(okHttpResponse,
        apolloResponse, Collections.<Record>emptyList());
  }
}

File: apollo-integration/src/test/kotlin/com/apollographql/apollo3/ApolloInterceptorChainTest.kt
package com.apollographql.apollo3

import com.apollographql.apollo3.Utils.immediateExecutor
import com.apollographql.apollo3.api.Input
import com.apollographql.apollo3.api.Response.Companion.builder
import com.apollographql.apollo3.exception.ApolloException
import com.apollographql.apollo3.exception.ApolloGenericException
import com.apollographql.apollo3.integration.normalizer.EpisodeHeroNameQuery
import com.apollographql.apollo3.integration.normalizer.type.Episode
import com.apollographql.apollo3.interceptor.ApolloInterceptor
import com.apollographql.apollo3.interceptor.ApolloInterceptor.*
import com.apollographql.apollo3.interceptor.ApolloInterceptorChain
import com.apollographql.apollo3.internal.interceptor.RealApolloInterceptorChain
import com.google.common.truth.Truth
import com.google.common.truth.Truth.assertThat
import okhttp3.*
import org.junit.Assert
import org.junit.Test
import java.util.concurrent.Executor
import java.util.concurrent.TimeoutException
import java.util.concurrent.atomic.AtomicInteger

class ApolloInterceptorChainTest {
  @Test
  @Throws(TimeoutException::class, InterruptedException::class)
  fun onProceedAsyncCalled_chainPassesControlToInterceptor() {
    val counter = AtomicInteger(1)
    val query = createQuery()
    val interceptor: ApolloInterceptor = object : ApolloInterceptor {
      override fun interceptAsync(request: InterceptorRequest, chain: ApolloInterceptorChain,
                                  dispatcher: Executor, callBack: CallBack) {
        counter.decrementAndGet()
      }

      override fun dispose() {}
    }
    val interceptors = listOf(interceptor)
    val chain = RealApolloInterceptorChain(interceptors)
    chain.proceedAsync(
        InterceptorRequest.builder(query).fetchFromCache(false).build(), immediateExecutor(),
        object : CallBack {
          override fun onResponse(response: InterceptorResponse) {}
          override fun onFailure(e: ApolloException) {}
          override fun onCompleted() {}
          override fun onFetch(sourceType: FetchSourceType) {}
        })

    //If counter's count doesn't go down to zero, it means interceptor's interceptAsync wasn't called
    //which means the test should fail.
    if (counter.get() != 0) {
      Assert.fail("Control not passed to the interceptor")
    }
  }

  @Test
  @Throws(TimeoutException::class, InterruptedException::class)
  fun onProceedAsyncCalled_correctInterceptorResponseIsReceived() {
    val counter = AtomicInteger(1)
    val query = createQuery()
    val expectedResponse = prepareInterceptorResponse(query)
    val interceptor: ApolloInterceptor = object : ApolloInterceptor {
      override fun interceptAsync(request: InterceptorRequest, chain: ApolloInterceptorChain,
                                  dispatcher: Executor, callBack: CallBack) {
        dispatcher.execute { callBack.onResponse(expectedResponse) }
      }

      override fun dispose() {}
    }
    val interceptors = listOf(interceptor)
    val chain = RealApolloInterceptorChain(interceptors)
    chain.proceedAsync(InterceptorRequest.builder(query).fetchFromCache(false).build(),
        immediateExecutor(), object : CallBack {
      override fun onResponse(response: InterceptorResponse) {
        Truth.assertThat(response).isEqualTo(expectedResponse)
        counter.decrementAndGet()
      }

      override fun onFailure(e: ApolloException) {}
      override fun onCompleted() {}
      override fun onFetch(sourceType: FetchSourceType) {}
    })
    if (counter.get() != 0) {
      Assert.fail("Interceptor's response not received")
    }
  }

  @Test
  @Throws(TimeoutException::class, InterruptedException::class)
  fun onProceedAsyncCalled_correctExceptionIsCaught() {
    val counter = AtomicInteger(1)
    val message = "ApolloException"
    val query = createQuery()
    val interceptor: ApolloInterceptor = object : ApolloInterceptor {
      override fun interceptAsync(request: InterceptorRequest, chain: ApolloInterceptorChain,
                                  dispatcher: Executor, callBack: CallBack) {
        dispatcher.execute {
          val apolloException = ApolloGenericException(message)
          callBack.onFailure(apolloException)
        }
      }

      override fun dispose() {}
    }
    val interceptors = listOf(interceptor)
    val chain = RealApolloInterceptorChain(interceptors)
    chain.proceedAsync(InterceptorRequest.builder(query).fetchFromCache(false).build(),
        immediateExecutor(), object : CallBack {
      override fun onResponse(response: InterceptorResponse) {}
      override fun onFailure(e: ApolloException) {
        assertThat(e.message).isEqualTo(message)
        counter.decrementAndGet()
      }

      override fun onCompleted() {}
      override fun onFetch(sourceType: FetchSourceType) {}
    })
    if (counter.get() != 0) {
      Assert.fail("Exception thrown by Interceptor not caught")
    }
  }

  @Test
  fun onDisposeCalled_interceptorIsDisposed() {
    val counter = AtomicInteger(1)
    val interceptor: ApolloInterceptor = object : ApolloInterceptor {
      override fun interceptAsync(request: InterceptorRequest, chain: ApolloInterceptorChain, dispatcher: Executor, callBack: CallBack) {}
      override fun dispose() {
        counter.decrementAndGet()
      }
    }
    val interceptors = listOf(interceptor)
    val chain = RealApolloInterceptorChain(interceptors)
    chain.dispose()
    if (counter.get() != 0) {
      Assert.fail("Interceptor's dispose method not called")
    }
  }

  private fun createQuery(): EpisodeHeroNameQuery {
    return EpisodeHeroNameQuery(episode = Input.present(Episode.EMPIRE))
  }

  private fun prepareInterceptorResponse(query: EpisodeHeroNameQuery): InterceptorResponse {
    val request = Request.Builder()
        .url("https://localhost:8080/")
        .build()
    val okHttpResponse = Response.Builder()
        .request(request)
        .protocol(Protocol.HTTP_2)
        .code(200)
        .message("Intercepted")
        .body(ResponseBody.create(MediaType.parse("text/plain; charset=utf-8"), "fakeResponse"))
        .build()
    val apolloResponse = builder<EpisodeHeroNameQuery.Data>(query).build()
    return InterceptorResponse(okHttpResponse, apolloResponse)
  }
}
